SIPS/sip-11.md

```diff
@@ -0,0 +1,68 @@
+---
+sip: 11
+title: Transaction insight severity levels
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/101
+author: Hassan Malik (@hmalik88)
+created: 2023-07-16
+---
+
+## Abstract
+
+This SIP proposes a way for snaps to provide extra friction for executing potentially dangerous transactions within MetaMask with the addition of a `severity` field to the already existing transaction insights API outlined in SIP-3.
+
+## Motivation
+
+One of the biggest issues with wallet users is a loss of funds through executing some sketchy or seemingly innocuous transaction. With the benefit of the transaction insights API, a snap can already run analysis on an unsigned transaction payload. By adding `severity` to the existing API, we allow for the snap to provide extra friction to the user if it determines that a transaction is malicious. Users often click through things without reading, by allowing for a warning we can add friction at points that we really think a user should have a second glance.
+
+This warning could be injected at any point where there are insights provided.
+
+### MetaMask Integration
+
+The `severity` key is added to the return object to indicate the severity level of the content being returned to the extension. This will help trigger certain UI in the extension. Currently, a warning modal will be triggered for content with a severity level of `critical`. The modal will require a checkbox to be checked before the user can continue with the transaction.
+
+In future SIPs, the `SeverityLevel` enum can be expanded to include other levels that can be also be used to influence the UI in the extension.
+
+Transaction insight snaps were previously triggered on view of their respective tabs, but with the addition of the `severity` key, execution would become unprompted in order to determine if a modal needs to be displayed as you reach the confirmation screen.
+
+### Snap Implementation
+
+The following is an example implementation of the API:
+
+```typescript
+import { OnTransactionHandler } from "@metamask/snap-types";
+
+export const onTransaction: OnTransactionHandler = async ({
+  transaction,
+  chainId,
+}) => {
+  const content = /* Get UI component with insights */;
+  const isContentCritical = /* Boolean checking if content is critical */
+  return isContentCritical ? { content, severity: 'critical' } : { content };
+};
+```
+
+The interface for the return value of an `onTransaction` export is:
+
+```typescript
+enum SeverityLevel {
+  Critical = 'critical',
```

Comment:
Just thinking out loud... So, it is mentioned:
```
In future SIPs, the `SeverityLevel` enum can be expanded to include other levels that can be also be used to influence the UI in the extension.
```
I'm thinking if we can proactively define some other severity levels? It would be somehow a proof that this is required to be built in a dynamic extensive way which has multiple enum values, etc. Otherwise, this could be just a boolean `critical?: boolean`. I'm not suggesting that we use a boolean, I can sense already that this will expand, just maybe if we expand it now, it will be more clear about the direction in which this is going and allow the other use cases from the beginning.

For example, how I see it:
```typescript
enum SeverityLevel {
  Critical = 'critical',
  Danger = 'danger',
  Suspicious = 'suspicious',
}
```
Just an idea to think about.

Reply:
From a user's perspective there is no difference between a "suspicious" transaction and a "critical" transaction. In both cases the user (likely) shouldn't go through with the transaction. In the future we may want to handle these cases different in the UI, but for the initial implementation, it will be just critical or nothing.